internal/overlay/overlay.go

```diff
@@ -66,16 +69,52 @@ func Apply(schema string, overlayFile string, w io.Writer) error {
 		return fmt.Errorf("failed to apply overlay to spec file %q: %w", specFile, err)
 	}
 
-	// Decode into an interface{} to avoid preserving the original formatting of the input file
-	// otherwise this can result in 1 line YAML files
-	var unformattedYaml interface{}
-	ys.Decode(&unformattedYaml)
+	bytes, err := render(ys, utils.HasYAMLExt(schema), yamlOut)
+	if err != nil {
+		return fmt.Errorf("failed to render document: %w", err)
+	}
 
-	enc := yaml.NewEncoder(w)
-	enc.SetIndent(2)
-	if err := enc.Encode(unformattedYaml); err != nil {
-		return fmt.Errorf("failed to encode spec file %q: %w", specFile, err)
+	if _, err := w.Write(bytes); err != nil {
+		return fmt.Errorf("failed to write to output: %w", err)
 	}
 
 	return nil
 }
+
+func render(y *yaml.Node, yamlIn bool, yamlOut bool) ([]byte, error) {
+	if yamlIn && yamlOut {
+		var res bytes.Buffer
+		if err := yaml.NewEncoder(&res).Encode(y); err != nil {
+			return nil, fmt.Errorf("failed to encode YAML: %w", err)
+		}
+		return res.Bytes(), nil
+	}
+
+	// Preserves key ordering
+	specBytes, err := json.YAMLNodeToJSON(y, "  ")
+	if err != nil {
+		return nil, fmt.Errorf("failed to convert YAML to JSON: %w", err)
+	}
+
+	if yamlOut {
+		// Use libopenapi to convert JSON to YAML to preserve key ordering
+		doc, err := libopenapi.NewDocument(specBytes)
```

Comment:
Would this break if applying an overlay to a file with external references?

Reply:
I'm not sure why it would, but let me add a test for that 👍

Reply:
Added tests for that. There was in fact an issue (needed docPath 🤦), but wanted to make sure you aren't expecting this to cause issues with the output file itself